### DISCUSSION TITLE
Automatic detection for `ext_modules`

### DISCUSSION THREAD

**Discussion by abravalheri** [pypa/setuptools#4569] [Dev]
In https://github.com/pypa/setuptools/issues/2220#issuecomment-2241150503, @jaraco suggested:

> I'd also like to consider some automatic detection such that static configuration is unnecessary for many cases (e.g. each C file in the package represents a module, or a particular arrangement of C files can represent a module).

What I would like to discuss in this thread is how this automatic detection would look like and in which conditions it would be triggered or skipped.

Something that we have to be mindful and potentially avoid are false negatives.

**Comment by abravalheri** [comment]
I feel like if we just scan the directories for `.c` files and consider that each `.c` file corresponds to an extension module we might find too many false positives.

I wonder if we could have a well defined directory suffix convention. For example:

```
.
├── pyproject.toml
└── foo.ext/
    ├── bar.c
    ├── baz.c
    └── foo.h
```

This could be translated to:

```python
Extension(
    name="foo",
    sources=["foo.ext/bar.c", "foo.ext/baz.c"],
    depends=["foo.ext/foo.h"],
)
```

**Comment by jaraco** [comment]
I like that!

A couple of other options:

Have static config in the .c/.h files indicate the relevant module:

```C
/* bar.c */
/* python-ext: foo */

...
```

Or, let there be a stub module with a specified extension/naming scheme to represent the module:

```python
# foo.cext
sources=['lib1/bar.c', 'lib2/baz.c']
include=['lib1/foo.h']
```

Overall, I'm liking your option best, but I thought I'd share these for consideration.
